// ng-sidebarjs/bundles/ng-sidebarjs.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core'], factory) :
	(factory((global.ngSidebarjs = {}),global.ng.core));
}(this, (function (exports,core) { 'use strict';

var commonjsGlobal = typeof window !== 'undefined' ? window : typeof global !== 'undefined' ? global : typeof self !== 'undefined' ? self : {};



function unwrapExports (x) {
	return x && x.__esModule && Object.prototype.hasOwnProperty.call(x, 'default') ? x['default'] : x;
}

function createCommonjsModule(fn, module) {
	return module = { exports: {} }, fn(module, module.exports), module.exports;
}

var sidebarjs = createCommonjsModule(function (module, exports) {
/*
 * SidebarJS
 * Version 4.0.0
 * https://github.com/SidebarJS/sidebarjs#readme
 */

(function (global, factory) {
	factory(exports);
}(commonjsGlobal, (function (exports) { function unwrapExports$$1 (x) {
	return x && x.__esModule && Object.prototype.hasOwnProperty.call(x, 'default') ? x['default'] : x;
}

function createCommonjsModule$$1(fn, module) {
	return module = { exports: {} }, fn(module, module.exports), module.exports;
}

var sidebarElement = createCommonjsModule$$1(function (module, exports) {
Object.defineProperty(exports, "__esModule", { value: true });
var sidebarjs = 'sidebarjs';
var isVisible = sidebarjs + "--is-visible";
var isMoving = sidebarjs + "--is-moving";
var LEFT_POSITION = 'left';
var RIGHT_POSITION = 'right';
var TRANSITION_DURATION = 400;
var POSITIONS = [LEFT_POSITION, RIGHT_POSITION];
var SidebarElement = /** @class */ (function () {
    function SidebarElement(config) {
        if (config === void 0) { config = {}; }
        var component = config.component, container = config.container, backdrop = config.backdrop, _a = config.documentMinSwipeX, documentMinSwipeX = _a === void 0 ? 10 : _a, _b = config.documentSwipeRange, documentSwipeRange = _b === void 0 ? 40 : _b, nativeSwipe = config.nativeSwipe, nativeSwipeOpen = config.nativeSwipeOpen, _c = config.position, position = _c === void 0 ? 'left' : _c, _d = config.backdropOpacity, backdropOpacity = _d === void 0 ? 0.3 : _d;
        this.component = component || document.querySelector("[" + sidebarjs + "]");
        this.container = container || SidebarElement.create(sidebarjs + "-container");
        this.backdrop = backdrop || SidebarElement.create(sidebarjs + "-backdrop");
        this.documentMinSwipeX = documentMinSwipeX;
        this.documentSwipeRange = documentSwipeRange;
        this.nativeSwipe = nativeSwipe !== false;
        this.nativeSwipeOpen = nativeSwipeOpen !== false;
        this.backdropOpacity = backdropOpacity;
        this.backdropOpacityRatio = 1 / backdropOpacity;
        var hasAllConfigDOMElements = component && container && backdrop;
        if (!hasAllConfigDOMElements) {
            try {
                this.transcludeContent();
            }
            catch (e) {
                throw new Error('You must define an element with [sidebarjs] attribute');
            }
        }
        if (this.nativeSwipe) {
            this.addNativeGestures();
            if (this.nativeSwipeOpen) {
                this.addNativeOpenGestures();
            }
        }
        this.setPosition(position);
        this.addAttrsEventsListeners(this.component.getAttribute(sidebarjs));
        this.backdrop.addEventListener('click', this.close.bind(this));
    }
    SidebarElement.prototype.toggle = function () {
        this.component.classList.contains(isVisible) ? this.close() : this.open();
    };
    SidebarElement.prototype.open = function () {
        this.component.classList.add(isVisible);
        this.setBackdropOpacity(this.backdropOpacity);
    };
    SidebarElement.prototype.close = function () {
        this.component.classList.remove(isVisible);
        this.backdrop.removeAttribute('style');
    };
    SidebarElement.prototype.isVisible = function () {
        return this.component.classList.contains(isVisible);
    };
    SidebarElement.prototype.setPosition = function (position) {
        var _this = this;
        this.component.classList.add(isMoving);
        this.position = POSITIONS.indexOf(position) >= 0 ? position : LEFT_POSITION;
        POSITIONS.forEach(function (POS) { return _this.component.classList.remove(sidebarjs + "--" + POS); });
        this.component.classList.add(sidebarjs + "--" + (this.hasRightPosition() ? RIGHT_POSITION : LEFT_POSITION));
        setTimeout(function () { return _this.component.classList.remove(isMoving); }, TRANSITION_DURATION);
    };
    SidebarElement.prototype.addAttrsEventsListeners = function (sidebarName) {
        var actions = ['toggle', 'open', 'close'];
        for (var i = 0; i < actions.length; i++) {
            var elements = document.querySelectorAll("[" + sidebarjs + "-" + actions[i] + "=\"" + sidebarName + "\"]");
            for (var j = 0; j < elements.length; j++) {
                if (!SidebarElement.elemHasListener(elements[j])) {
                    elements[j].addEventListener('click', this[actions[i]].bind(this));
                    SidebarElement.elemHasListener(elements[j], true);
                }
            }
        }
    };
    SidebarElement.prototype.hasLeftPosition = function () {
        return this.position === LEFT_POSITION;
    };
    SidebarElement.prototype.hasRightPosition = function () {
        return this.position === RIGHT_POSITION;
    };
    SidebarElement.prototype.transcludeContent = function () {
        this.container.innerHTML = this.component.innerHTML;
        this.component.innerHTML = '';
        this.component.appendChild(this.container);
        this.component.appendChild(this.backdrop);
    };
    SidebarElement.prototype.addNativeGestures = function () {
        this.component.addEventListener('touchstart', this.onTouchStart.bind(this));
        this.component.addEventListener('touchmove', this.onTouchMove.bind(this));
        this.component.addEventListener('touchend', this.onTouchEnd.bind(this));
    };
    SidebarElement.prototype.addNativeOpenGestures = function () {
        document.addEventListener('touchstart', this.onSwipeOpenStart.bind(this));
        document.addEventListener('touchmove', this.onSwipeOpenMove.bind(this));
        document.addEventListener('touchend', this.onSwipeOpenEnd.bind(this));
    };
    SidebarElement.prototype.onTouchStart = function (e) {
        this.initialTouch = e.touches[0].pageX;
    };
    SidebarElement.prototype.onTouchMove = function (e) {
        var documentSwiped = this.initialTouch - e.touches[0].clientX;
        var sidebarMovement = this.getSidebarPosition(documentSwiped);
        this.touchMoveSidebar = -documentSwiped;
        if (sidebarMovement <= this.container.clientWidth) {
            this.moveSidebar(this.touchMoveSidebar);
        }
    };
    SidebarElement.prototype.onTouchEnd = function () {
        this.component.classList.remove(isMoving);
        this.container.removeAttribute('style');
        this.backdrop.removeAttribute('style');
        Math.abs(this.touchMoveSidebar) > (this.container.clientWidth / 3.5) ? this.close() : this.open();
        delete this.initialTouch;
        delete this.touchMoveSidebar;
    };
    SidebarElement.prototype.moveSidebar = function (movement) {
        this.component.classList.add(isMoving);
        SidebarElement.vendorify(this.container, 'transform', "translate(" + movement + "px, 0)");
        this.updateBackdropOpacity(movement);
    };
    SidebarElement.prototype.updateBackdropOpacity = function (movement) {
        var swipeProgress = 1 - (Math.abs(movement) / this.container.clientWidth);
        var opacity = swipeProgress / this.backdropOpacityRatio;
        this.setBackdropOpacity(opacity);
    };
    SidebarElement.prototype.setBackdropOpacity = function (opacity) {
        this.backdrop.style.opacity = opacity.toString();
    };
    SidebarElement.prototype.onSwipeOpenStart = function (e) {
        if (this.targetElementIsBackdrop(e)) {
            return;
        }
        var clientWidth = document.body.clientWidth;
        var touchPositionX = e.touches[0].clientX;
        var documentTouch = this.hasLeftPosition() ? touchPositionX : clientWidth - touchPositionX;
        if (documentTouch < this.documentSwipeRange) {
            this.onTouchStart(e);
        }
    };
    SidebarElement.prototype.onSwipeOpenMove = function (e) {
        if (!this.targetElementIsBackdrop(e) && this.initialTouch && !this.isVisible()) {
            var documentSwiped = e.touches[0].clientX - this.initialTouch;
            var sidebarMovement = this.getSidebarPosition(documentSwiped);
            if (sidebarMovement > 0) {
                SidebarElement.vendorify(this.component, 'transform', 'translate(0, 0)');
                SidebarElement.vendorify(this.component, 'transition', 'none');
                this.openMovement = sidebarMovement * (this.hasLeftPosition() ? -1 : 1);
                this.moveSidebar(this.openMovement);
            }
        }
    };
    SidebarElement.prototype.onSwipeOpenEnd = function () {
        if (this.openMovement) {
            delete this.openMovement;
            this.component.removeAttribute('style');
            this.onTouchEnd();
        }
    };
    SidebarElement.prototype.getSidebarPosition = function (swiped) {
        return (this.container.clientWidth - (this.hasLeftPosition() ? swiped : -swiped));
    };
    SidebarElement.prototype.targetElementIsBackdrop = function (e) {
        var touchedElement = e.target;
        return touchedElement.hasAttribute(sidebarjs + "-backdrop");
    };
    SidebarElement.create = function (element) {
        var el = document.createElement('div');
        el.setAttribute(element, '');
        return el;
    };
    SidebarElement.vendorify = function (el, prop, val) {
        var Prop = prop.charAt(0).toUpperCase() + prop.slice(1);
        var prefs = ['Moz', 'Webkit', 'O', 'ms'];
        el.style[prop] = val;
        for (var i = 0; i < prefs.length; i++) {
            el.style[prefs[i] + Prop] = val;
        }
        return el;
    };
    SidebarElement.elemHasListener = function (elem, value) {
        return elem && (value === true || value === false) ? elem.sidebarjsListener = value : !!elem.sidebarjsListener;
    };
    return SidebarElement;
}());
exports.SidebarElement = SidebarElement;

});

unwrapExports$$1(sidebarElement);
var sidebarElement_1 = sidebarElement.SidebarElement;

var sidebarService = createCommonjsModule$$1(function (module, exports) {
Object.defineProperty(exports, "__esModule", { value: true });

var SidebarService = /** @class */ (function () {
    function SidebarService() {
        this.instances = {};
    }
    SidebarService.prototype.create = function (options) {
        if (options === void 0) { options = {}; }
        var name = options.component && options.component.getAttribute('sidebarjs') || '';
        this.instances[name] = new sidebarElement.SidebarElement(options);
        return this.instances[name];
    };
    SidebarService.prototype.open = function (sidebarName) {
        if (sidebarName === void 0) { sidebarName = ''; }
        if (this.instances[sidebarName]) {
            this.instances[sidebarName].open();
        }
    };
    SidebarService.prototype.close = function (sidebarName) {
        if (sidebarName === void 0) { sidebarName = ''; }
        if (this.instances[sidebarName]) {
            this.instances[sidebarName].close();
        }
    };
    SidebarService.prototype.toggle = function (sidebarName) {
        if (sidebarName === void 0) { sidebarName = ''; }
        if (this.instances[sidebarName]) {
            this.instances[sidebarName].toggle();
        }
    };
    SidebarService.prototype.isVisible = function (sidebarName) {
        if (sidebarName === void 0) { sidebarName = ''; }
        return !!this.instances[sidebarName] && this.instances[sidebarName].isVisible();
    };
    SidebarService.prototype.setPosition = function (position, sidebarName) {
        if (sidebarName === void 0) { sidebarName = ''; }
        if (this.instances[sidebarName]) {
            this.instances[sidebarName].setPosition(position);
        }
    };
    SidebarService.prototype.elemHasListener = function (elem, value) {
        return sidebarElement.SidebarElement.elemHasListener(elem, value);
    };
    SidebarService.prototype.destroy = function (sidebarName) {
        if (sidebarName === void 0) { sidebarName = ''; }
        if (this.instances[sidebarName]) {
            delete this.instances[sidebarName];
        }
    };
    return SidebarService;
}());
exports.SidebarService = SidebarService;

});

unwrapExports$$1(sidebarService);
var sidebarService_1 = sidebarService.SidebarService;

var src = createCommonjsModule$$1(function (module, exports) {
Object.defineProperty(exports, "__esModule", { value: true });

exports.SidebarElement = sidebarElement.SidebarElement;

exports.SidebarService = sidebarService.SidebarService;

});

var index = unwrapExports$$1(src);
var src_1 = src.SidebarElement;
var src_2 = src.SidebarService;

exports['default'] = index;
exports.SidebarElement = src_1;
exports.SidebarService = src_2;

Object.defineProperty(exports, '__esModule', { value: true });

})));
});

unwrapExports(sidebarjs);

var sidebarjs$2 = sidebarjs;

var sidebarjs_1 = sidebarjs$2.SidebarElement;
var sidebarjs_2 = sidebarjs$2.SidebarService;

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var SidebarJSComponent = /** @class */ (function () {
    function SidebarJSComponent(sidebarService, renderer) {
        this.sidebarService = sidebarService;
        this.renderer = renderer;
    }
    /**
     * @return {?}
     */
    SidebarJSComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ configDomElements = this.defineConfigDomElements();
        this.setSidebarAttributes(this.sidebarjsName, configDomElements);
        this.sidebarService.create(Object.assign({}, this.sidebarjsConfig, configDomElements));
    };
    /**
     * @return {?}
     */
    SidebarJSComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.sidebarService.destroy(this.sidebarjsName);
    };
    /**
     * @return {?}
     */
    SidebarJSComponent.prototype.defineConfigDomElements = /**
     * @return {?}
     */
    function () {
        return {
            component: this.component.nativeElement,
            container: this.container.nativeElement,
            backdrop: this.backdrop.nativeElement,
        };
    };
    /**
     * @param {?=} sidebarName
     * @param {?=} configDomElements
     * @return {?}
     */
    SidebarJSComponent.prototype.setSidebarAttributes = /**
     * @param {?=} sidebarName
     * @param {?=} configDomElements
     * @return {?}
     */
    function (sidebarName, configDomElements) {
        if (sidebarName === void 0) { sidebarName = ''; }
        this.renderer.setAttribute(configDomElements.component, 'sidebarjs', sidebarName);
        this.renderer.setAttribute(configDomElements.container, 'sidebarjs-container', '');
        this.renderer.setAttribute(configDomElements.backdrop, 'sidebarjs-backdrop', '');
    };
    SidebarJSComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'sidebar-js',
                    template: "<div #component> <div #container><ng-content></ng-content></div> <div #backdrop></div> </div> ",
                    styles: ["[sidebarjs-backdrop], [sidebarjs] { top: 0; left: 0; width: 100%; height: 100%; } [sidebarjs].sidebarjs--left, [sidebarjs].sidebarjs--left [sidebarjs-container] { -webkit-transform: translate(-100%, 0); transform: translate(-100%, 0); } [sidebarjs].sidebarjs--right, [sidebarjs].sidebarjs--right [sidebarjs-container] { -webkit-transform: translate(100%, 0); transform: translate(100%, 0); } [sidebarjs].sidebarjs--left [sidebarjs-container] { -webkit-box-shadow: 2px 0 4px rgba(0, 0, 0, 0.2); box-shadow: 2px 0 4px rgba(0, 0, 0, 0.2); } [sidebarjs].sidebarjs--right [sidebarjs-container] { -webkit-box-shadow: -2px 0 4px rgba(0, 0, 0, 0.2); box-shadow: -2px 0 4px rgba(0, 0, 0, 0.2); margin-left: auto; } [sidebarjs-backdrop] { position: absolute; background: #000; opacity: 0; -webkit-transition: opacity 0.3s ease; transition: opacity 0.3s ease; } [sidebarjs-container] { position: relative; z-index: 1; display: -webkit-box; display: -ms-flexbox; display: flex; -webkit-box-orient: vertical; -webkit-box-direction: normal; -ms-flex-direction: column; flex-direction: column; width: 90%; max-width: 300px; height: 100%; background: #fff; -webkit-transition: -webkit-transform ease 0.3s; transition: -webkit-transform ease 0.3s; transition: transform ease 0.3s; transition: transform ease 0.3s, -webkit-transform ease 0.3s; } [sidebarjs] { position: fixed; z-index: 9999; -webkit-transition: -webkit-transform 0s ease 0.3s; transition: -webkit-transform 0s ease 0.3s; transition: transform 0s ease 0.3s; transition: transform 0s ease 0.3s, -webkit-transform 0s ease 0.3s; } [sidebarjs].sidebarjs--is-visible { -webkit-transform: translate(0, 0); transform: translate(0, 0); -webkit-transition: -webkit-transform 0s ease 0s; transition: -webkit-transform 0s ease 0s; transition: transform 0s ease 0s; transition: transform 0s ease 0s, -webkit-transform 0s ease 0s; } [sidebarjs].sidebarjs--is-visible [sidebarjs-container] { -webkit-transform: translate(0, 0); transform: translate(0, 0); } [sidebarjs].sidebarjs--is-moving { -webkit-transition: none; transition: none; } [sidebarjs].sidebarjs--is-moving [sidebarjs-container], [sidebarjs].sidebarjs--is-moving [sidebarjs-backdrop] { -webkit-transition: none; transition: none; } "]
                },] },
    ];
    /** @nocollapse */
    SidebarJSComponent.ctorParameters = function () { return [
        { type: sidebarjs_2, },
        { type: core.Renderer2, },
    ]; };
    SidebarJSComponent.propDecorators = {
        "sidebarjsName": [{ type: core.Input },],
        "sidebarjsConfig": [{ type: core.Input },],
        "component": [{ type: core.ViewChild, args: ['component',] },],
        "container": [{ type: core.ViewChild, args: ['container',] },],
        "backdrop": [{ type: core.ViewChild, args: ['backdrop',] },],
    };
    return SidebarJSComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var SidebarJSToggleDirective = /** @class */ (function () {
    function SidebarJSToggleDirective(sidebarService) {
        this.sidebarService = sidebarService;
    }
    /**
     * @param {?} element
     * @return {?}
     */
    SidebarJSToggleDirective.prototype.toggle = /**
     * @param {?} element
     * @return {?}
     */
    function (element) {
        if (!this.sidebarService.elemHasListener(element)) {
            this.sidebarService.toggle(this.sidebarjsToggle);
        }
    };
    SidebarJSToggleDirective.decorators = [
        { type: core.Directive, args: [{
                    selector: '[sidebarjsToggle]'
                },] },
    ];
    /** @nocollapse */
    SidebarJSToggleDirective.ctorParameters = function () { return [
        { type: sidebarjs_2, },
    ]; };
    SidebarJSToggleDirective.propDecorators = {
        "sidebarjsToggle": [{ type: core.Input },],
        "toggle": [{ type: core.HostListener, args: ['click', ['$event.target'],] },],
    };
    return SidebarJSToggleDirective;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var SidebarJSOpenDirective = /** @class */ (function () {
    function SidebarJSOpenDirective(sidebarService) {
        this.sidebarService = sidebarService;
    }
    /**
     * @param {?} element
     * @return {?}
     */
    SidebarJSOpenDirective.prototype.open = /**
     * @param {?} element
     * @return {?}
     */
    function (element) {
        if (!this.sidebarService.elemHasListener(element)) {
            this.sidebarService.open(this.sidebarjsOpen);
        }
    };
    SidebarJSOpenDirective.decorators = [
        { type: core.Directive, args: [{
                    selector: '[sidebarjsOpen]'
                },] },
    ];
    /** @nocollapse */
    SidebarJSOpenDirective.ctorParameters = function () { return [
        { type: sidebarjs_2, },
    ]; };
    SidebarJSOpenDirective.propDecorators = {
        "sidebarjsOpen": [{ type: core.Input },],
        "open": [{ type: core.HostListener, args: ['click', ['$event.target'],] },],
    };
    return SidebarJSOpenDirective;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var SidebarJSCloseDirective = /** @class */ (function () {
    function SidebarJSCloseDirective(sidebarService) {
        this.sidebarService = sidebarService;
    }
    /**
     * @param {?} element
     * @return {?}
     */
    SidebarJSCloseDirective.prototype.close = /**
     * @param {?} element
     * @return {?}
     */
    function (element) {
        if (!this.sidebarService.elemHasListener(element)) {
            this.sidebarService.close(this.sidebarjsClose);
        }
    };
    SidebarJSCloseDirective.decorators = [
        { type: core.Directive, args: [{
                    selector: '[sidebarjsClose]'
                },] },
    ];
    /** @nocollapse */
    SidebarJSCloseDirective.ctorParameters = function () { return [
        { type: sidebarjs_2, },
    ]; };
    SidebarJSCloseDirective.propDecorators = {
        "sidebarjsClose": [{ type: core.Input },],
        "close": [{ type: core.HostListener, args: ['click', ['$event.target'],] },],
    };
    return SidebarJSCloseDirective;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var COMPONENTS = [
    SidebarJSComponent,
    SidebarJSToggleDirective,
    SidebarJSOpenDirective,
    SidebarJSCloseDirective,
];
var SidebarJSModule = /** @class */ (function () {
    function SidebarJSModule() {
    }
    /**
     * @return {?}
     */
    SidebarJSModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: SidebarJSModule,
            providers: [sidebarjs_2],
        };
    };
    SidebarJSModule.decorators = [
        { type: core.NgModule, args: [{
                    declarations: COMPONENTS.slice(),
                    exports: COMPONENTS.slice()
                },] },
    ];
    /** @nocollapse */
    SidebarJSModule.ctorParameters = function () { return []; };
    return SidebarJSModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */

exports.SidebarJSComponent = SidebarJSComponent;
exports.SidebarJSService = sidebarjs_2;
exports.SidebarJSModule = SidebarJSModule;
exports.SidebarJSToggleDirective = SidebarJSToggleDirective;
exports.SidebarJSOpenDirective = SidebarJSOpenDirective;
exports.SidebarJSCloseDirective = SidebarJSCloseDirective;

Object.defineProperty(exports, '__esModule', { value: true });

})));
